src/endpoints/pool/pool.service.ts

```diff
@@ -0,0 +1,113 @@
+
+@Injectable()
+export class PoolService {
+  constructor(
+    private readonly gatewayService: GatewayService,
+    private readonly apiConfigService: ApiConfigService,
+    private readonly cacheService: CacheService,
+  ) { }
+
+  async getTransactionFromPool(txHash: string): Promise<TransactionInPool | undefined> {
+    const pool = await this.getEntirePool();
+    const transaction = pool.find(tx => tx.txHash === txHash);
+
+    return transaction;
+  }
+
+  async getPoolCount(filter: PoolFilter): Promise<number> {
+    const pool = await this.getEntirePool();
+    return this.applyFilters(pool, filter).length;
+  }
+
+  async getPool(
+    queryPagination: QueryPagination,
+    filter: PoolFilter,
+  ): Promise<TransactionInPool[]> {
+    if (!this.apiConfigService.isTransactionPoolEnabled()) {
+      return [];
+    }
+
+    const { from, size } = queryPagination;
+    const entirePool = await this.getEntirePool();
+    const pool = this.applyFilters(entirePool, filter);
+    return pool.slice(from, from + size);
+  }
+
+  async getEntirePool(): Promise<TransactionInPool[]> {
+    return await this.cacheService.getOrSet(
+      CacheInfo.TransactionPool.key,
+      async () => await this.getTxPoolRaw(),
+      CacheInfo.TransactionPool.ttl
+    );
+  }
+
+  async getTxPoolRaw(): Promise<TransactionInPool[]> {
+    const pool = await this.gatewayService.getTransactionPool();
+    return this.parseTransactions(pool);
+  }
+
+  private parseTransactions(rawPool: TxPoolGatewayResponse): TransactionInPool[] {
+    const transactionPool: TransactionInPool[] = [];
+    if (rawPool && rawPool.txPool) {
+      transactionPool.push(...this.processTransactionType(rawPool.txPool.regularTransactions || [], TransactionType.Transaction));
+      transactionPool.push(...this.processTransactionType(rawPool.txPool.smartContractResults || [], TransactionType.SmartContractResult));
+      transactionPool.push(...this.processTransactionType(rawPool.txPool.rewards || [], TransactionType.Reward));
+    }
+
+    return transactionPool;
+  }
+
+  private processTransactionType(transactions: any[], transactionType: TransactionType): TransactionInPool[] {
+    return transactions.map(tx => this.parseTransaction(tx.txFields, transactionType));
+  }
+
+  private parseTransaction(tx: TxInPoolFields, type: TransactionType): TransactionInPool {
+    return new TransactionInPool({
+      txHash: tx.hash || '',
+      sender: tx.sender || '',
+      receiver: tx.receiver || '',
+      nonce: tx.nonce || 0,
+      value: tx.value || '',
+      gasPrice: tx.gasprice || 0,
+      gasLimit: tx.gaslimit || 0,
+      data: tx.data || '',
+      type: type || TransactionType.Transaction,
+    });
+  }
+
+  private applyFilters(pool: TransactionInPool[], filters: PoolFilter): TransactionInPool[] {
+    if (!filters.receiver && !filters.sender && !filters.type) {
+      return pool;
+    }
+
+    const results: TransactionInPool[] = [];
+
+    for (const transaction of pool) {
```

Comment:
instead of using the for and continue strategy, we can use the `.filter` function for every filter parameter. in this way, we can skip the initial check which might cause bugs when adding a new filter

Reply:
refactored by using the filter function